### PR TITLE
Reducing Parameters of Mask GCNN

### PR DESCRIPTION
**Pull request by chrisrothUT**
This PR does two things.

1. Allows masks to be passed through the GCNN constructor 
2. Reduces the parameters in the module to include only the "unmasked" parameters. This is useful for `QGTJacobianPyTree`, where you need `n_parameters*n_samples` to fit into memory 

Here's an example of a GCNN with 2nd nearest neighbor convolutions on the square lattice:

```
#3 x 3 convolutional filter
input_mask = np.zeros([3,3])
for i in range(-1,2):
  for j in range(-1,2):
    input_mask[i][j] = 1
input_mask = input_mask.ravel()

# repeat mask over point group for hidden mask
hidden_mask = np.repeat(np.expand_dims(input_mask,1),repeats=8,axis=1).ravel()

g = nk.graph.Square(8)
ma = GCNN(symmetries=g,parity=1,input_mask=input_mask,hidden_mask=hidden_mask...)
```

I'd like to make this a bit more user friendly if possible

### CODE REVIEW

**Comment by PhilipVinc** [comment]
I'm busy today, I'll give a look tomorrow evening.  From a quick look I gave yesterday I can say
 
Part 1, the part of the PR that allows to use masking together with GCNN looks ok (minus proper docstrings and a couple tests.) Could you properly write in a formula of the docstring of the publicGCNN function how the masks act on the weights? what input and hidden mask do to the calculation? Other than that it looks ok to me

Our Symmetric layers already support masking. You want to change how this is supported internally to completely remove the masked indices. I agree this lowers the memory cost. 

I had played with this last year in another project for some autoregressive networks, and found (not so explainably to me) that this yielded much higher computational cost for the Jacobian. Could you show me a benchmark or two showing that this masking does not cause increases in the computational cost wrt the previous scheme?
Also, @wdphy16 can you give me your opinion about this change? have you ever played with this?

Note: this reduces the parameter count, so it breaks parameter loading. Parameter loading/shape is not part of our stable API and I'm completely in favour of changing it if it makes sense, but I remember @attila-i-szabo wanted to add a loading shim the last time we changed the shape... Do we want it this time as well? 
In any case, if we change the format we will need big warning boxes in the docstrings saying that in version 3.7 we changed the format.

**Comment by attila-i-szabo** [comment]
> > I had played with this last year in another project for some autoregressive networks, and found (not so explainably to me) that this yielded much higher computational cost for the Jacobian. Could you show me a benchmark or two showing that this masking does not cause increases in the computational cost wrt the previous scheme?
> 
> I added the benchmark in "masked_layers.py". As it turns out masking the layers like this creates a pretty big compile time overhead. It looks like there's a 20x compile time overhead at least in the small model I tested. After the code is compiled it basically runs the same with and without a mask.

I don't think it's a time cost necessarily. We also noticed this issue that the backpropagation effectively calculates the gradient of all possible kernel entries and then discards the masked ones. That's why we required #1347, so the calculation fits into memory. In production code, we should look at how to define a custom gradient that doesn't do this, or whether that's possible. 

> Note: this reduces the parameter count, so it breaks parameter loading. Parameter loading/shape is not part of our stable API and I'm completely in favour of changing it if it makes sense, but I remember @attila-i-szabo wanted to add a loading shim the last time we changed the shape... Do we want it this time as well?
> In any case, if we change the format we will need big warning boxes in the docstrings saying that in version 3.7 we changed the format.

I don't think it was my insistence as much as @PhilipVinc's. I don't mind either way, but it will be a lot more work to do the conversion routines for this change than it was for adding extra dimensions here and there...

**Comment by PhilipVinc** [comment]
>I don't think it's a time cost necessarily. We also noticed this issue that the backpropagation effectively calculates the gradient of all possible kernel entries and then discards the masked ones. That's why we required https://github.com/netket/netket/pull/1347, so the calculation fits into memory. In production code, we should look at how to define a custom gradient that doesn't do this, or whether that's possible.

Yeah, that's how back propagation would implement this... 
in 'backward' pass you do all operations in the opposite order.
So if in the forwad pass you start from "few parameters" -> "Build large matrix with the few parameters inside" -> "matmul with other data" -> reductions.... 
You can see that in the backward pass you will be carrying around the gradients of the "big matrix with few parameters inside" all the way until the last step when you discard all except those that matter.

The only way to avoid this would be for you to write the operation "matmul with other data of a big matrix with few parameters inside" and define a custom adjoint for it that never concretises the big matrix, assuming it's possible.
It might also be worth looking into sparse matrices? jax now supports them (experimentally) and they are now at a decent point... `jax.experimental.sparse`. I don't know if it makes sense for this use case though...

> I don't think it was my insistence as much as @PhilipVinc's. I don't mind either way, but it will be a lot more work to do the conversion routines for this change than it was for adding extra dimensions here and there

sorry, I might have misremembered. Then it's all good. 
I don't think we need conversion routines... unless you guys want them. 

Ps: congrats on the paper

**Comment by attila-i-szabo** [comment]
> Yeah, that's how back propagation would implement this... in 'backward' pass you do all operations in the opposite order. So if in the forwad pass you start from "few parameters" -> "Build large matrix with the few parameters inside" -> "matmul with other data" -> reductions.... You can see that in the backward pass you will be carrying around the gradients of the "big matrix with few parameters inside" all the way until the last step when you discard all except those that matter.
> 
> The only way to avoid this would be for you to write the operation "matmul with other data of a big matrix with few parameters inside" and define a custom adjoint for it that never concretises the big matrix, assuming it's possible. It might also be worth looking into sparse matrices? jax now supports them (experimentally) and they are now at a decent point... `jax.experimental.sparse`. I don't know if it makes sense for this use case though...

It's a bit more delicate than this. IIRC from my experience, there is no such problem with the `irrep` implementation, only the `fft` one. It seems that computing all that stuff is not strictly necessary, but FFT is too opaque for JAX to eliminate the extra calculations when jitting. I think it'd be important to think through how to write a custom adjoint that doesn't waste memory, but I don't have the time/energy to do that :( @chrisrothUT, I think I analysed this with you a few months back, maybe you want to look up whether we had any practical insights then...

> Ps: congrats on the paper

Thank you!

**Comment by PhilipVinc** [comment]
> It seems that computing all that stuff is not strictly necessary, but FFT is too opaque for JAX to eliminate the extra calculations when jitting

Hmm.. ok. I'm not sure that I understand. But would be interesting to look into it in the future..

File: netket/nn/symmetric_linear.py [hunk 1]
```diff
@@ -661,8 +700,8 @@ def DenseSymm(symmetries, point_group=None, mode="auto", shape=None, **kwargs):
         features: The number of output features. The full output shape
             is :code:`[n_batch,features,n_symm]`.
         use_bias: A bool specifying whether to add a bias to the output (default: True).
-        mask: An optional array of shape [n_sites] consisting of ones and zeros
-            that can be used to give the kernel a particular shape.
+        input_mask: Optional array of shape [n_sites] that used to restrict the convolutional
```
[the remark this answers is not in the document]
Comment (PhilipVinc) [reply netket/nn/symmetric_linear.py:703]: proposal: why not calling this `mask` instead of `input mask`? 
For the DenseSymm there is no other mask so no need to discern the two... or am I wrong?

File: netket/models/equivariant.py [hunk 2]
```diff
@@ -652,10 +680,18 @@ def GCNN(
         bias_init: Initializer for the biases of all layers.
         complex_output: If True, ensures that the network output is always complex.
             Necessary when network parameters are real but some `characters` are negative.
-
+        input_mask: Optional array of shape [n_sites] that used to restrict the convolutional
```
[the remark this answers is not in the document]
Comment (PhilipVinc) [reply netket/models/equivariant.py:683]: does this need to be an explicit input? can't we automatically compute it?

File: netket/models/equivariant.py [hunk 3]
```diff
@@ -287,6 +299,10 @@ class GCNN_Parity_FFT(nn.Module):
     """The nonlinear activation function between hidden layers."""
     output_activation: Any = identity
     """The nonlinear activation before the output."""
+    input_mask: Array = None
+    """mask for input to hidden weights"""
+    hidden_mask: Array = None
+    """mask for hidden to hidden weights"""
```
Comment (PhilipVinc) [inline netket/models/equivariant.py:305]: add the shape to the docstring please

File: netket/nn/symmetric_linear.py [hunk 4]
```diff
@@ -749,8 +788,8 @@ def DenseEquivariant(
         features: The number of output features. The full output shape
             is [n_batch,features,n_symm].
         use_bias: A bool specifying whether to add a bias to the output (default: True).
-        mask: An optional array of shape [n_sites] consisting of ones and zeros
-            that can be used to give the kernel a particular shape.
+        input_mask: Optional array of shape [n_symm] that used to restrict the convolutional
```
[the remark this answers is not in the document]
Comment (chrisrothUT) [reply netket/nn/symmetric_linear.py:791]: Yeah we can just call it mask in the layer

[the remark this answers is not in the document]
Comment (PhilipVinc) [reply netket/nn/symmetric_linear.py:791]: great, can you rename it then?

File: netket/nn/symmetric_linear.py [hunk 5]
```diff
@@ -334,15 +341,26 @@ def __call__(self, x: Array) -> Array:
...
-            self.kernel_init,
-            (self.features, in_features, self.n_point * self.n_cells),
-            self.param_dtype,
-        )
-
         if self.mask is not None:
-            kernel = kernel * jnp.expand_dims(self.scaled_mask, (0, 1))
+            kernel_params = self.param(
+                "kernel",
+                self.kernel_init,
```
[the remark this answers is not in the document]
Comment (PhilipVinc) [reply netket/nn/symmetric_linear.py:347]: right! indeed

**Comment by chrisrothUT** [comment]
> Ps: congrats on the paper

Thanks!

Think I've made all of the changes you requested! How do I get rid of the flake warning again?

**Comment by PhilipVinc** [comment]
>  How do I get rid of the flake warning again?

the list is https://github.com/netket/netket/actions/runs/3490968435/jobs/5843055582 
```python
netket/models/equivariant.py:81:89: W291 trailing whitespace
[13](https://github.com/netket/netket/actions/runs/3490968435/jobs/5843055582#step:5:14)
netket/models/equivariant.py:84:92: W291 trailing whitespace
[14](https://github.com/netket/netket/actions/runs/3490968435/jobs/5843055582#step:5:15)
netket/models/equivariant.py:85:101: W291 trailing whitespace
[15](https://github.com/netket/netket/actions/runs/3490968435/jobs/5843055582#step:5:16)
netket/models/equivariant.py:204:89: W291 trailing whitespace
[16](https://github.com/netket/netket/actions/runs/3490968435/jobs/5843055582#step:5:17)
netket/models/equivariant.py:207:92: W291 trailing whitespace
[17](https://github.com/netket/netket/actions/runs/3490968435/jobs/5843055582#step:5:18)
netket/models/equivariant.py:208:101: W291 trailing whitespace
[18](https://github.com/netket/netket/actions/runs/3490968435/jobs/5843055582#step:5:19)
netket/models/equivariant.py:312:89: W291 trailing whitespace
[19](https://github.com/netket/netket/actions/runs/3490968435/jobs/5843055582#step:5:20)
netket/models/equivariant.py:315:92: W291 trailing whitespace
[20](https://github.com/netket/netket/actions/runs/3490968435/jobs/5843055582#step:5:21)
netket/models/equivariant.py:316:101: W291 trailing whitespace
[21](https://github.com/netket/netket/actions/runs/3490968435/jobs/5843055582#step:5:22)
netket/models/equivariant.py:499:89: W291 trailing whitespace
[22](https://github.com/netket/netket/actions/runs/3490968435/jobs/5843055582#step:5:23)
netket/models/equivariant.py:502:92: W291 trailing whitespace
[23](https://github.com/netket/netket/actions/runs/3490968435/jobs/5843055582#step:5:24)
netket/models/equivariant.py:503:101: W291 trailing whitespace
[24](https://github.com/netket/netket/actions/runs/3490968435/jobs/5843055582#step:5:25)
netket/nn/symmetric_linear.py:61:89: W291 trailing whitespace
[25](https://github.com/netket/netket/actions/runs/3490968435/jobs/5843055582#step:5:26)
netket/nn/symmetric_linear.py:129:9: F841 local variable 'dtype' is assigned to but never used
[26](https://github.com/netket/netket/actions/runs/3490968435/jobs/5843055582#step:5:27)
netket/nn/symmetric_linear.py:166:89: W291 trailing whitespace
[27](https://github.com/netket/netket/actions/runs/3490968435/jobs/5843055582#step:5:28)
netket/nn/symmetric_linear.py:296:92: W291 trailing whitespace
[28](https://github.com/netket/netket/actions/runs/3490968435/jobs/5843055582#step:5:29)
netket/nn/symmetric_linear.py:297:101: W291 trailing whitespace
[29](https://github.com/netket/netket/actions/runs/3490968435/jobs/5843055582#step:5:30)
netket/nn/symmetric_linear.py:434:92: W291 trailing whitespace
[30](https://github.com/netket/netket/actions/runs/3490968435/jobs/5843055582#step:5:31)
netket/nn/symmetric_linear.py:435:101: W291 trailing whitespace
[31](https://github.com/netket/netket/actions/runs/3490968435/jobs/5843055582#step:5:32)
netket/nn/symmetric_linear.py:599:92: W291 trailing whitespace
[32](https://github.com/netket/netket/actions/runs/3490968435/jobs/5843055582#step:5:33)
netket/nn/symmetric_linear.py:600:101: W291 trailing whitespace
[33](https://github.com/netket/netket/actions/runs/3490968435/jobs/5843055582#step:5:34)
```

(by the way, you can run it locally by running `flake8 netket`)


It's simply a few complaints about trailing whitespaces at the end of some lines (probably in docstrings), and it also complaints about that dtype variable you never use.

File: netket/nn/symmetric_linear.py [hunk 6]
```diff
@@ -605,10 +646,8 @@ def __call__(self, x: Array) -> Array:
         else:
             bias = None
 
-        if self.mask is not None:
-            kernel = kernel * jnp.expand_dims(self.scaled_mask, (0, 1))
-
         kernel, bias, x = promote_dtype(kernel, bias, x, dtype=None)
+        dtype = x.dtype
```
[the remark this answers is not in the document]
Comment (PhilipVinc) [reply netket/nn/symmetric_linear.py]: @chrisrothUT  ah, I think you wanted this to be
```python
        kernel, bias, x = promote_dtype(kernel, bias, x, dtype=x.dtype)
```

maybe?

**Comment by PhilipVinc** [comment]
yes, that's the idea. There's still a few (and that spurious dtype remains) https://github.com/netket/netket/actions/runs/3490968435/jobs/5843055582

Also could you add two lines to the changelog? one in `new features` saying the new options and one in `breaking changes` saying that we are breaking the storage format for those networks if they have a mask.

**Comment by chrisrothUT** [comment]
I added the changes to the Changelog. As for the whitespaces, when I run flake8 on my computer I don't get any errors (at least in the 'GCNN' part of the code). Also, I manually checked for the whitespaces and they are gone so I'm not sure what the issue is.

**Comment by PhilipVinc** [comment]
looks good now, it seems you fixed the whitespace issues

**Review by PhilipVinc — APPROVED** [review]

**Comment by PhilipVinc** [comment]
For posterity: the test that is failing here is not relevant to this PR, and has been flaky for a while  For reasons I don't understand.

But this time, it finally threw an intelligible stack trace that I paste here so not to loose it...

```python
6723
Extension modules: jaxlib.cpu_feature_guard, numpy.core._multiarray_umath, numpy.core._multiarray_tests, numpy.linalg._umath_linalg, numpy.fft._pocketfft_internal, numpy.random._common, numpy.random.bit_generator, numpy.random._bounded_integers, numpy.random._mt19937, numpy.random.mtrand, numpy.random._philox, numpy.random._pcg64, numpy.random._sfc64, numpy.random._generatorfree(): corrupted unsorted chunks
[6724](https://github.com/netket/netket/actions/runs/3559379372/jobs/5978693072#step:8:6725)
Fatal Python error: Aborted
[6725](https://github.com/netket/netket/actions/runs/3559379372/jobs/5978693072#step:8:6726)
```

to be investigated..

**Comment by PhilipVinc** [comment]
thanks a lot @chrisrothUT !
